backend/pkg/api/data_access/vdb_management.go

```diff
@@ -836,83 +835,68 @@ func (d *DataAccessService) GetValidatorDashboardGroupExists(ctx context.Context
 }
 
 func (d *DataAccessService) AddValidatorDashboardValidators(ctx context.Context, dashboardId t.VDBIdPrimary, groupId uint64, validators []t.VDBValidator) ([]t.VDBPostValidatorsData, error) {
+	result := []t.VDBPostValidatorsData{}
+
 	if len(validators) == 0 {
 		// No validators to add
 		return nil, nil
 	}
 
-	pubkeys := []struct {
-		ValidatorIndex t.VDBValidator `db:"validatorindex"`
-		Pubkey         []byte         `db:"pubkey"`
-	}{}
-
-	addedValidators := []struct {
-		ValidatorIndex t.VDBValidator `db:"validator_index"`
-		GroupId        uint64         `db:"group_id"`
-	}{}
-
-	// Query to find the pubkey for each validator index
-	pubkeysQuery := `
-		SELECT
-			validatorindex,
-			pubkey
-		FROM validators
-		WHERE validatorindex = ANY($1)
-	`
-
-	// Query to add the validators to the dashboard and group
-	addValidatorsQuery := `
-		INSERT INTO users_val_dashboards_validators (dashboard_id, group_id, validator_index)
-			VALUES
-	`
-
-	for idx := range validators {
-		addValidatorsQuery += fmt.Sprintf("($1, $2, $%d), ", idx+3)
-	}
-	addValidatorsQuery = addValidatorsQuery[:len(addValidatorsQuery)-2] // remove trailing comma
-
-	// If a validator is already in the dashboard, update the group
-	// If the validator is already in that group nothing changes but we will include it in the result anyway
-	addValidatorsQuery += `
-		ON CONFLICT (dashboard_id, validator_index) DO UPDATE SET
-			dashboard_id = EXCLUDED.dashboard_id,
-			group_id = EXCLUDED.group_id,
-			validator_index = EXCLUDED.validator_index
-		RETURNING validator_index, group_id
-	`
-
-	// Find all the pubkeys
-	err := d.alloyReader.SelectContext(ctx, &pubkeys, pubkeysQuery, pq.Array(validators))
+	tx, err := d.userWriter.BeginTxx(ctx, nil)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("error starting db transactions to insert validators for a dashboard: %w", err)
 	}
+	defer utils.Rollback(tx)
 
-	// Add all the validators to the dashboard and group
-	addValidatorsArgsIntf := []interface{}{dashboardId, groupId}
-	for _, validatorIndex := range validators {
-		addValidatorsArgsIntf = append(addValidatorsArgsIntf, validatorIndex)
-	}
-	err = d.alloyWriter.SelectContext(ctx, &addedValidators, addValidatorsQuery, addValidatorsArgsIntf...)
-	if err != nil {
-		return nil, err
-	}
+	numArgs := 3
+	batchSize := 65535 / numArgs // max 65535 params per batch, since postgres uses int16 for binding input params
+	batchIdx, allIdx := 0, 0
```

Comment:
goqu has a built in batch insert feature, see https://github.com/gobitfly/beaconchain/blob/staging/backend/pkg/notification/queuing.go#L448 for its usage

Reply:
As discussed the linked code part does not avoid the parameter limit.

The only way I have found is e.g. [here](https://github.com/gobitfly/beaconchain/blob/743a86528b0930da74e63d6a5420dfe5a86bd368/backend/pkg/exporter/modules/ethstore.go#L190) with the use of `pgx` and `CopyFrom`/`CopyFromSlice`. If a high email notification count has to work for the linked part you could use that. I just can't use it here because it does not support `ON CONFLICT` handling.